float64 linear_e
float64 angle_e